cmd/soroban-rpc/internal/test/upgrade_test.go

```diff
@@ -0,0 +1,128 @@
+package test
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/creachadair/jrpc2"
+	"github.com/creachadair/jrpc2/jhttp"
+	"github.com/stellar/go/keypair"
+	"github.com/stellar/go/txnbuild"
+	"github.com/stellar/go/xdr"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/stellar/soroban-rpc/cmd/soroban-rpc/internal/db"
+)
+
+func TestUpgradeFrom20To21(t *testing.T) {
+	if GetCoreMaxSupportedProtocol() != 21 {
+		t.Skip("Only test this for protocol 21")
+	}
+	test := NewTest(t, &TestConfig{
+		ProtocolVersion: 20,
+	})
+
+	ch := jhttp.NewChannel(test.sorobanRPCURL(), nil)
+	client := jrpc2.NewClient(ch, nil)
+
+	sourceAccount := keypair.Root(StandaloneNetworkPassphrase)
+	address := sourceAccount.Address()
+	account := txnbuild.NewSimpleAccount(address, 0)
+
+	helloWorldContract := getHelloWorldContract(t)
+
+	params := preflightTransactionParams(t, client, txnbuild.TransactionParams{
+		SourceAccount:        &account,
+		IncrementSequenceNum: true,
+		Operations: []txnbuild.Operation{
+			createInstallContractCodeOperation(account.AccountID, helloWorldContract),
+		},
+		BaseFee: txnbuild.MinBaseFee,
+		Preconditions: txnbuild.Preconditions{
+			TimeBounds: txnbuild.NewInfiniteTimeout(),
+		},
+	})
+
+	tx, err := txnbuild.NewTransaction(params)
+	assert.NoError(t, err)
+	sendSuccessfulTransaction(t, client, sourceAccount, tx)
+
+	// Upgrade to protocol 21 and re-upload the contract, which should cause a caching of the contract
+	// estimations
+	test.UpgradeProtocol(21)
+	// Wait for the ledger to advance, so that the simulation library passes the right protocol
+	rpcDB := test.daemon.GetDB()
+	initialLedgerSequence, err := db.NewLedgerEntryReader(rpcDB).GetLatestLedgerSequence(context.Background())
+	require.NoError(t, err)
+	newLedgerSequence := uint32(0)
+	for i := 0; i < 60; i++ {
+		time.Sleep(time.Second)
+		newLedgerSequence, err = db.NewLedgerEntryReader(rpcDB).GetLatestLedgerSequence(context.Background())
+		require.NoError(t, err)
+		if newLedgerSequence > initialLedgerSequence {
+			break
+		}
+	}
+	require.Greater(t, newLedgerSequence, initialLedgerSequence, "rpc didn't start ingesting protocol 21")
```

Comment:
could remove the `for loop` in favor of `require.Eventually(t, func() {...}, time.Minute, time.Second)`

Reply:
Nice, I will use this!

Reply:
We should probably use this in the state expiration tests as well